source/fem-dem/fluid_dynamics_vans_matrix_free_operators.cc

```diff
@@ -0,0 +1,500 @@
+          // ν(∇v,ɛ∇δu)
+          gradient_result[i] =
+            this->kinematic_viscosity * vf_value * gradient[i];
+          // ν(v,∇ɛ∇δu)
+          value_result[i] =
+            this->kinematic_viscosity * vf_gradient * gradient[i];
+          // -(∇·v,ɛδp)
+          gradient_result[i][i] += -vf_value * value[dim];
+          // -(v,δp∇ɛ)
+          value_result[i] += -vf_gradient[i] * value[dim];
+          // +(q,ɛ∇·δu)
+          value_result[dim] += vf_value * gradient[i][i];
+          // +(q,∇ɛ·δu)
+          value_result[dim] += vf_gradient[i] * value[i];
+
+          for (unsigned int k = 0; k < dim; ++k)
+            {
+              // +(v,ɛ(u·∇)δu + ɛ(δu·∇)u)
+              value_result[i] +=
+                vf_value * (gradient[i][k] * previous_values[k] +
+                            previous_gradient[i][k] * value[k]);
+            }
+          // +(v,ɛ ∂t δu)
+          if (transient)
+            value_result[i] += vf_value * (*bdf_coefs)[0] * value[i];
+        }
+
+      // PSPG Jacobian
+      for (unsigned int i = 0; i < dim; ++i)
+        {
+          for (unsigned int k = 0; k < dim; ++k)
+            {
+              // (-νɛ∆δu + ɛ(u·∇)δu + ɛ(δu·∇)u)·τ∇q
+              gradient_result[dim][i] +=
+                tau * vf_value *
+                (-this->kinematic_viscosity * hessian_diagonal[i][k] +
+                 gradient[i][k] * previous_values[k] +
+                 previous_gradient[i][k] * value[k]);
+            }
+          // +ɛ(∂t δu)·τ∇q
+          if (transient)
+            gradient_result[dim][i] +=
+              tau * vf_value * (*bdf_coefs)[0] * value[i];
+        }
+      // (ɛ∇δp)τ·∇q
+      gradient_result[dim] += tau * vf_value * gradient[dim];
+
+      // Grad-div stabilization Jacobian
+      // (∇·v,γ(ɛ∇·δu+δu·∇ɛ))
+      if (cfd_dem_parameters.grad_div == true)
+        {
+          for (unsigned int i = 0; i < dim; ++i)
+            {
+              for (unsigned int k = 0; k < dim; ++k)
+                {
+                  gradient_result[i][i] +=
+                    grad_div_gamma(cell, q) *
+                    (vf_value * gradient[k][k] + value[k] * vf_gradient[k]);
+                }
+            }
+        }
+
+      // SUPG Jacobian
+      for (unsigned int i = 0; i < dim; ++i)
+        {
+          for (unsigned int k = 0; k < dim; ++k)
+            {
+              // Part 1
+              for (unsigned int l = 0; l < dim; ++l)
+                {
+                  // +(ɛ(u·∇)δu + ɛ(δu·∇)u - νɛ∆δu)τ(u·∇)v
+                  gradient_result[i][k] +=
+                    tau * vf_value * previous_values[k] *
+                    (gradient[i][l] * previous_values[l] +
+                     vf_value * previous_gradient[i][l] * value[l] -
+                     this->kinematic_viscosity * hessian_diagonal[i][l]);
+                }
+              // +(ɛ∇δp)τ(u·∇)v
+              gradient_result[i][k] +=
+                tau * vf_value * previous_values[k] * (gradient[dim][i]);
```

Comment:
Is gradient correct here or should it be previous?

Reply:
No it is gradient because we are assembling the jacobian matrix. See line 245 (it is delta p here and not p).